// @antv/g6-plugin/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Bundling", {
  enumerable: true,
  get: function get() {
    return _bundling.default;
  }
});
Object.defineProperty(exports, "EdgeFilterLens", {
  enumerable: true,
  get: function get() {
    return _edgeFilterLens.default;
  }
});
Object.defineProperty(exports, "Fisheye", {
  enumerable: true,
  get: function get() {
    return _fisheye.default;
  }
});
Object.defineProperty(exports, "Grid", {
  enumerable: true,
  get: function get() {
    return _grid.default;
  }
});
Object.defineProperty(exports, "ImageMinimap", {
  enumerable: true,
  get: function get() {
    return _imageMinimap.default;
  }
});
Object.defineProperty(exports, "Legend", {
  enumerable: true,
  get: function get() {
    return _legend.default;
  }
});
Object.defineProperty(exports, "Menu", {
  enumerable: true,
  get: function get() {
    return _menu.default;
  }
});
Object.defineProperty(exports, "Minimap", {
  enumerable: true,
  get: function get() {
    return _minimap.default;
  }
});
Object.defineProperty(exports, "PluginBase", {
  enumerable: true,
  get: function get() {
    return _base.default;
  }
});
Object.defineProperty(exports, "SnapLine", {
  enumerable: true,
  get: function get() {
    return _snapline.default;
  }
});
Object.defineProperty(exports, "TimeBar", {
  enumerable: true,
  get: function get() {
    return _timeBar.default;
  }
});
Object.defineProperty(exports, "ToolBar", {
  enumerable: true,
  get: function get() {
    return _toolBar.default;
  }
});
Object.defineProperty(exports, "Tooltip", {
  enumerable: true,
  get: function get() {
    return _tooltip.default;
  }
});
exports.default = void 0;

var _grid = _interopRequireDefault(require("./grid"));

var _menu = _interopRequireDefault(require("./menu"));

var _minimap = _interopRequireDefault(require("./minimap"));

var _bundling = _interopRequireDefault(require("./bundling"));

var _fisheye = _interopRequireDefault(require("./fisheye"));

var _toolBar = _interopRequireDefault(require("./toolBar"));

var _tooltip = _interopRequireDefault(require("./tooltip"));

var _timeBar = _interopRequireDefault(require("./timeBar"));

var _imageMinimap = _interopRequireDefault(require("./imageMinimap"));

var _edgeFilterLens = _interopRequireDefault(require("./edgeFilterLens"));

var _snapline = _interopRequireDefault(require("./snapline"));

var _base = _interopRequireDefault(require("./base"));

var _legend = _interopRequireDefault(require("./legend"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Plugin = {
  PluginBase: _base.default,
  Menu: _menu.default,
  Grid: _grid.default,
  Minimap: _minimap.default,
  Bundling: _bundling.default,
  ToolBar: _toolBar.default,
  Tooltip: _tooltip.default,
  Fisheye: _fisheye.default,
  TimeBar: _timeBar.default,
  ImageMinimap: _imageMinimap.default,
  EdgeFilterLens: _edgeFilterLens.default,
  SnapLine: _snapline.default,
  Legend: _legend.default
};
var _default = Plugin;
exports.default = _default;
