// @antv/g6-core/lib/index.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  registerNode: true,
  registerCombo: true,
  registerEdge: true,
  registerBehavior: true,
  BaseGlobal: true,
  AbstractGraph: true,
  Shape: true,
  Arrow: true,
  Marker: true,
  Util: true,
  AbstractLayout: true,
  AbstractEvent: true,
  Node: true,
  Edge: true,
  Hull: true,
  Combo: true
};
Object.defineProperty(exports, "AbstractEvent", {
  enumerable: true,
  get: function get() {
    return _event.default;
  }
});
Object.defineProperty(exports, "AbstractGraph", {
  enumerable: true,
  get: function get() {
    return _graph.default;
  }
});
Object.defineProperty(exports, "AbstractLayout", {
  enumerable: true,
  get: function get() {
    return _layout.default;
  }
});
Object.defineProperty(exports, "Arrow", {
  enumerable: true,
  get: function get() {
    return _element.Arrow;
  }
});
exports.BaseGlobal = void 0;
Object.defineProperty(exports, "Combo", {
  enumerable: true,
  get: function get() {
    return _combo.default;
  }
});
Object.defineProperty(exports, "Edge", {
  enumerable: true,
  get: function get() {
    return _edge.default;
  }
});
Object.defineProperty(exports, "Hull", {
  enumerable: true,
  get: function get() {
    return _hull.default;
  }
});
Object.defineProperty(exports, "Marker", {
  enumerable: true,
  get: function get() {
    return _element.Marker;
  }
});
Object.defineProperty(exports, "Node", {
  enumerable: true,
  get: function get() {
    return _node.default;
  }
});
Object.defineProperty(exports, "Shape", {
  enumerable: true,
  get: function get() {
    return _element.default;
  }
});
Object.defineProperty(exports, "Util", {
  enumerable: true,
  get: function get() {
    return _util.default;
  }
});
exports.registerNode = exports.registerEdge = exports.registerCombo = exports.registerBehavior = exports.default = void 0;

var _behavior = _interopRequireDefault(require("./behavior"));

var _graph = _interopRequireDefault(require("./graph/graph"));

var _element = _interopRequireWildcard(require("./element"));

var _global = _interopRequireDefault(require("./global"));

var _util = _interopRequireDefault(require("./util"));

var _layout = _interopRequireDefault(require("./graph/controller/layout"));

var _event = _interopRequireDefault(require("./graph/controller/event"));

var _node = _interopRequireDefault(require("./item/node"));

var _edge = _interopRequireDefault(require("./item/edge"));

var _hull = _interopRequireDefault(require("./item/hull"));

var _combo = _interopRequireDefault(require("./item/combo"));

var _types = require("./types");

Object.keys(_types).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _types[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _types[key];
    }
  });
});

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// 用于 PC 和 Mobile 端分别实现 layout 和 updateLayoutCfg 方法
var registerNode = _element.default.registerNode;
exports.registerNode = registerNode;
var registerEdge = _element.default.registerEdge;
exports.registerEdge = registerEdge;
var registerCombo = _element.default.registerCombo;
exports.registerCombo = registerCombo;
var registerBehavior = _behavior.default.registerBehavior;
exports.registerBehavior = registerBehavior;
var BaseGlobal = _global.default;
exports.BaseGlobal = BaseGlobal;
var _default = {
  version: _global.default.version,
  AbstractGraph: _graph.default,
  BaseGlobal: BaseGlobal,
  Util: _util.default,
  Shape: _element.default,
  Node: _node.default,
  Edge: _edge.default,
  Combo: _combo.default,
  Hull: _hull.default,
  registerNode: _element.default.registerNode,
  registerEdge: _element.default.registerEdge,
  registerCombo: _element.default.registerCombo,
  registerBehavior: _behavior.default.registerBehavior,
  Arrow: _element.Arrow,
  Marker: _element.Marker,
  AbstractLayout: _layout.default,
  AbstractEvent: _event.default
};
exports.default = _default;
