// @antv/g-webgpu/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Camera", {
  enumerable: true,
  get: function get() {
    return _Camera.Camera;
  }
});
Object.defineProperty(exports, "Geometry", {
  enumerable: true,
  get: function get() {
    return _geometry.Geometry;
  }
});
Object.defineProperty(exports, "Kernel", {
  enumerable: true,
  get: function get() {
    return _Kernel.Kernel;
  }
});
Object.defineProperty(exports, "Material", {
  enumerable: true,
  get: function get() {
    return _material.Material;
  }
});
Object.defineProperty(exports, "Renderable", {
  enumerable: true,
  get: function get() {
    return _Renderable.Renderable;
  }
});
Object.defineProperty(exports, "World", {
  enumerable: true,
  get: function get() {
    return _World.World;
  }
});

var _Camera = require("./camera/Camera");

var _geometry = require("./geometry");

var _Kernel = require("./Kernel");

var _material = require("./material");

var _Renderable = require("./renderable/Renderable");

var _World = require("./World");
